### ISSUE TITLE
operator-sdk extension is bound to Openshift only

### ISSUE THREAD

**Issue by mswiderski** [quarkiverse/quarkus-operator-sdk#10]
Looks like the operator-sdk is currently brining Openshift client always and by that makes it impossible to run in vanilla kubernetes. When run on kubernetes it fails with following error in the logs

````
2021-03-16 07:21:02,073 ERROR [io.jav.ope.Operator] (main) Not connected to any cluster. Exiting!
````

moreover it does not really say what happened and when debugging it given NPE on Openshift client calls. And having `System.exit` is not really container friendly :)
https://github.com/java-operator-sdk/java-operator-sdk/blob/master/operator-framework-core/src/main/java/io/javaoperatorsdk/operator/Operator.java#L53

Is this intended to make it OpenShift only extension or shall it be either configurable or having separate extensions for kubernetes or Openshift?

**Comment by metacosm** [comment]
This is already addressed in more recent versions as this was a bug in the kubernetes-client. The extension is definitely not OpenShift-only but it should allow you to access OpenShift-only resources if needed like Route, etc.

The `System.exit` call is a little harsh but it should only occur when a connection to the server cannot be established (which wasn't the case in this instance due to a bug in the OpenShift client: https://github.com/fabric8io/kubernetes-client/issues/2856. 

What would you say is a better behavior if the operator cannot connect to the cluster?

**Comment by mswiderski** [comment]
excellent to hear it was already addressed and hopefully it will be available in 1.13 of quarkus - the way I was able to workaround was with excludes of Openshift related ones and put kubernetes instead. Works but not desired ;)

> What would you say is a better behavior if the operator cannot connect to the cluster?

I'd say throwing an exception that would failed application at boot time would be better. The reason why I actually noticed the `System.exit` was when this error happened it simply made the container stale - kubernetes was still being it as alive but it was nothing anything. Having health checks most likely would make it spin constantly and fail again until health check kills it. 
Wdyt?

**Comment by metacosm** [comment]
> excellent to hear it was already addressed and hopefully it will be available in 1.13 of quarkus - the way I was able to workaround was with excludes of Openshift related ones and put kubernetes instead. Works but not desired ;)
>
Indeed. Maybe we should document this particular use case, though?

> > What would you say is a better behavior if the operator cannot connect to the cluster?
> 
> I'd say throwing an exception that would failed application at boot time would be better. The reason why I actually noticed the `System.exit` was when this error happened it simply made the container stale - kubernetes was still being it as alive but it was nothing anything. Having health checks most likely would make it spin constantly and fail again until health check kills it.
> Wdyt?

Ah yes, didn't think of that. I assumed that kubernetes would kill the pod in that case. I'll open an issue on the SDK to throw an exception instead.

**Comment by mswiderski** [comment]
> > excellent to hear it was already addressed and hopefully it will be available in 1.13 of quarkus - the way I was able to workaround was with excludes of Openshift related ones and put kubernetes instead. Works but not desired ;)
> 
> Indeed. Maybe we should document this particular use case, though?

I am not sure this requires documentation if the Openshift client will work out of the box. Alternatively I was thinking to have extension for Openshift and extension for kubernetes ... but if the Openshift client will work for both don't think it is needed. 

A side question, does the Openshift client brings many additional classes/dependencies? Mainly for native compilation if that might have any negative impact... I guess not but better to double check :)
> 
> > > What would you say is a better behavior if the operator cannot connect to the cluster?
> > 
> > 
> > I'd say throwing an exception that would failed application at boot time would be better. The reason why I actually noticed the `System.exit` was when this error happened it simply made the container stale - kubernetes was still being it as alive but it was nothing anything. Having health checks most likely would make it spin constantly and fail again until health check kills it.
> > Wdyt?
> 
> Ah yes, didn't think of that. I assumed that kubernetes would kill the pod in that case. I'll open an issue on the SDK to throw an exception instead.

thanks!

**Comment by metacosm** [comment]
> A side question, does the Openshift client brings many additional classes/dependencies? Mainly for native compilation if that might have any negative impact... I guess not but better to double check :)

That's also something I've been wondering about before making the change to use the OpenShift client as opposed to the plain k8s one. In my very limited testing, if you're not using any OpenShift-specific resources, it doesn't appear to have any impact on the native binary size (I guess because Substrate culls any unneeded code from your binary). But it's definitely something to keep an eye out for.

Can we close this issue if it's resolved to your satisfaction?

**Comment by mswiderski** [comment]
> Can we close this issue if it's resolved to your satisfaction?

yes, go ahead and close it. As soon as there is new release I will run my tests on it and hopefully exclusions won't be required anymore.

**Comment by metacosm** [comment]
I'll keep it open until then as incentive to release a new version! 😁

**Comment by mswiderski** [comment]
just tested it out with recently released quarkus 1.13.0.CR1 (that includes Kubernetes Client 5.2.1) and I can confirm that now it works without any excludes with vanilla kubernetes cluster.

**Comment by metacosm** [comment]
Fixed by 654972e51567a3e90bd21d28b701e317d596be2e
